### DISCUSSION TITLE
Using Gitextensions with .NET Core on Linux

### DISCUSSION THREAD

**Discussion by borzag** [gitextensions/gitextensions#9470] [General] [answered]
Hi,
Loves GE and have been using it for a couple of years. I'm developing in Linux currently and would really like to use it there.
Read some discussions and checkin comments about GE and .NET Core. I saw that some work has been done regarding that. What is the status? Is it possible to execute GE on some .NET Core version? Tried a little bit with some .NET Core versions, with the [Portable Version](https://github.com/gitextensions/gitextensions/releases/download/v3.5.2/GitExtensions-Portable-3.5.2.12435-77413699a.zip) but had no luck. Does there exist any special build that is possible to execute with some version of .NET Core?

Does it exist some install instruction to get it work on Linux?

**Comment by borzag** [comment, marked as answer]
Hi,

I will answer my question my self :-). 

I have realized that I have mixed up which technology that is available on which platform. GitExtensions is using WinForms for the GUI. It's possible to use WinForms in .NET Core 3.x and .NET 5. It's also possible to execute .NET Core 3.x and .NET 5 on Linux, but that does not apply to .NET applications using WinForms. This GUI technology works only on Windows in these platforms. ([See here form more information ](https://docs.microsoft.com/en-us/dotnet/desktop/winforms/migration/?view=netdesktop-5.0).

So the conclusion to my question is that GitExtensions can currently not be executed on Linux, while running on .NET Core 3.x and .NET 5 , and using WinForms. With [Mono](https://github.com/gitextensions/gitextensions/wiki/How-To:-run-Git-Extensions-on-Linux) it's still possible.

Thanks for a great application!

**Reply by gerhardol** [reply]
Some more details:
* 2.5x is Mono compatible. Not much have happened with Mono the last ten years or so, unfortunately a dead end and it was hindering development and would have killed the app completely if kept, unfortunately.
* 3.0-3.5 uses .NET 4.6
* master (likely 3.6 or 4.0) uses .NET5 that is available in Linux. WinForms is open source but not ported to other platforms (no active work that I have heard of).

It should be possible to use Wine or similar too.

There has been discussions to rewrite the UI layer to something that is cross platform, but there has not been any volunteers to to take a year or so off without wages to undertake this.

**Reply by RussKie** [reply]
> * master (likely 3.6 or 4.0) uses .NET5 that is available in Linux. WinForms is open source but not ported to other platforms (no active work that I have heard of).

There are no plans (that I am aware of) to port Windows Forms to any other platforms (hint: the name). It may be possible to run it under Wine ¯\\_(ツ)\_/¯

**Comment by arnonuem** [comment]
Maybe its a good idea to say goodbye to .Net and separate logic from ui and use some platform independent frameworks like electron or whatever.

**Reply by gerhardol** [reply]
If someone has a lot of time to spare (5000h? 10000h?) it would be an alternative.

**Reply by arnonuem** [reply]
How tightly is the current logic coupled to .NET UI components?

**Reply by gerhardol** [reply]
It is WinForms, mostly intertwined

**Reply by arnonuem** [reply]
Maybe it is a good idea to setup a ticket to put an ACL between winforms and the actual logic so that the ui becomes exchangable. That would be a good start for others than to contribute multiple UIs to the, like i call it, best git client out there. Unfortunately i am restricted to version 2.5 as many others out there. So all your development you put into 2.5++ is only i would guess for the minority of win users :/

**Comment by RussKie** [comment]
We keep having these discussions every few years ;)

We can consider Blazor as a x-plat vehicle, Avalonia was suggested as an
option as well.

We could start with FormCommit, but it will require re-arch and re-write of
a number of API. Which will require sizable effort, that the core team
can't facilitate.

**Reply by arnonuem** [reply]
But really you could start a poll to see who is working with linux vs windows. I think that the most effort should go into providing a ui which works on linux. So why not freeze new features and just concentrate on refactoring first?

**Reply by pmiossec** [reply]
>  So why not freeze new features and just concentrate on refactoring first?

I think this kind of remark is a lack of understanding on how an open-source project is working.

Every one is working on its free time to improve the project and bring the features and improvements he want to see in the application.

Asking that every contributors stop to do what brings value to them just because you want something done is not really realistic.

If someone want to see multi-platform support, the best is that he forks the project and do all the refactoring needed there.
If the project takes momentum, contributors will come and maybe features will be ported from this project to the fork.

And if everything went well, the current project will die because the fork superseded it.

Until it happens, you can't ask people to work on what they want (because that please them or they need it)

**Reply by gerhardol** [reply]
I really like cross platform. But Mono was getting out of date and there were very little feedback and contributions from Mono users and it was hindering development why Mono could not be supported. And rewriting is not really an option... But someone that is interested, could try make a concept.

New functionality could be added to 2.x too (the code base is not that different to 3.x) for someone interested.

**Reply by arnonuem** [reply]
It makes only sense to make a concept if one is familiar with the development and the history of the application.

**Comment by RussKie** [comment]
No. You welcome to fork and start a rewrite.
